feat(javascript): Refactor &  Compress Long

theweipeng opened the pull request:
## Refactor the code generator
1. Currently, JavaScript can only register a Object by Type.object, because the generator is too simple
2. Primitive generator of types like string、number can't be inlined in the holder collection, which cause necessary polymorphic that affect the performance

## Enhancement performance
1. Currently, Buffer.latin1Slice would transfer from native code generated by JIT to v8 runtime, This has a significant impact on performance, so we use String.fromCharCode to create little string when the string length is less than 15.  It is a magic number, but I have tested it and it works fine.

## Improve
When useSliceString is disabled, the performance of deserialization improves by 100%. It is 3 times faster than native JSON and twice as fast as protobuf.

When useSliceString is enabled, the performance of deserialization improves by 30%. It is 5 times faster than native JSON and 3 times faster than protobuf.

### Before:

#### Disable useSliceString
![image](https://github.com/apache/incubator-fury/assets/16490211/83278ece-0eba-4aa5-81a5-e2806bfd1997)

#### Enable useSliceString
![image](https://github.com/apache/incubator-fury/assets/16490211/abaeb420-274a-4aae-b432-750e96061668)

### After:
#### Disable useSliceString
![image](https://github.com/apache/incubator-fury/assets/16490211/2d0e1d3f-e3e4-432e-8b20-8d934b848a76)

#### Enable useSliceString
![image](https://github.com/apache/incubator-fury/assets/16490211/a9939cbf-cede-4a1d-8720-ac0f04a3abd6)

theweipeng commented:
@bytemain  Hi, could you please take a review, It remove the limit which mentioned in #1304

bytemain commented:
<img width="1149" alt="image" src="https://github.com/apache/incubator-fury/assets/13938334/5a43b6e1-523e-4145-82ed-5900f5ca1e6e">

I checked the snapshot and donot sure is this a problem?

bytemain commented:
fantasitic work, this is the thing I wanted of a serialization framework

theweipeng commented:
> it seem that this is not a problem, all the call will happen in this closure
> 
> <img alt="image" width="1149" src="https://private-user-images.githubusercontent.com/13938334/294660531-5a43b6e1-523e-4145-82ed-5900f5ca1e6e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ1MzEzMDcsIm5iZiI6MTcwNDUzMTAwNywicGF0aCI6Ii8xMzkzODMzNC8yOTQ2NjA1MzEtNWE0M2I2ZTEtNTIzZS00MTQ1LTgyZWQtNTkwMGY1Y2ExZTZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA2VDA4NTAwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiZjc2ZmJlNmJhYTQ5OWRmNTRkY2VhNzYxZTFhY2YxZTg4Yjk0OTE4MzdjZjczMGY5ZmIwZTgyYjFiZDI5NGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.D3ZFZNlAogNWEPaogXbBpI_jPsSwUvjKg_LRS-YTEnk">
> I checked the snapshot and donot sure is this a problem?
This is not a problem, but declare the readInner before the read is better. I will change it

bytemain commented on javascript/packages/fury/lib/classResolver.ts:
```diff
@@ -158,26 +184,15 @@ export default class SerializerResolver {
       const len = binaryReader.int16();
+      const start = binaryReader.getCursor();
       binaryReader.skip(len);
       this.readStringPool.push(Lazystring.fromPair(start, len));
```
I think the class name should use camel case

```diff
- class Lazystring
+ class LazyString
```

bytemain commented on javascript/packages/fury/lib/gen/index.ts:
```diff
@@ -0,0 +1,86 @@
+  return new Function(funcString);
+};
+
+function regDependences(fury: Fury, description: TypeDescription) {
```
```diff
- regDependences
+ regDependencies
```

bytemain commented on javascript/packages/fury/lib/type.ts:
```diff
@@ -28,7 +29,7 @@ export type BinaryReader = ReturnType<typeof BinaryReader>;
   STRING = 13,
   ARRAY = 25,
-  TUPLE = 25,
+  TUPLE = 25.1,
```
after the enum value change, we can remove the redundant code now:
<img width="1581" alt="image" src="https://github.com/apache/incubator-fury/assets/13938334/54a7cfd9-3541-43bc-82b3-b33207cfab32">

bytemain commented:
> @bytemain Hi, could you please take a review, It remove the limit which mentioned in #1304

I think you can add some test case in the #1304:

- https://github.com/apache/incubator-fury/pull/1304/files#diff-9d33aa5fce03357373ad7c54b9636452600284f3e7153064438e277158a3e9cfR60

bytemain commented:
Looks good to me, but has some typo.

And we can improve the style of generated code, this can be done in further

theweipeng commented:
> Looks good to me, but has some typo.
> 
> And we can improve the style of generated code, this can be done in further

I fixed it and added an afterCodeGenerated hook to allow users to format the code. Formatting is not a default action, as I believe we should keep things simple and avoid including too many dependencies.

chaokunyang commented on javascript/packages/fury/lib/gen/number.ts:
```diff
@@ -0,0 +1,105 @@
+    builder => builder.reader.float()
+  )
+);
+Register.reg(InternalSerializerType.DOUBLE,
```
```suggestion
Registry.register(InternalSerializerType.DOUBLE,
```

chaokunyang commented on javascript/packages/fury/lib/gen/router.ts:
```diff
@@ -17,28 +17,22 @@
-};
+type SerializerGeneratorConstructor = new (description: TypeDescription, builder: Builder, scope: Scope) => SerializerGenerator;
+
+export class Register {
```
This can be named as `Registry`, so `reg` can be renamed to `register`

chaokunyang replied:
Seems it's codegen related, maybe we can name it as `CodegenRegistry` or something like

chaokunyang replied:
In java we have a class named as `CodeGenerator`, and has a class named `CodecBuilder`. Hope it can give some inspiration

chaokunyang commented on javascript/packages/fury/lib/reader.ts:
```diff
@@ -144,21 +190,67 @@ export const BinaryReader = (config: Config) => {
+  }
+
+  function varUInt64() {
+    let byte_ = bigUInt8();
```
If remaing bytes is greater than 8, we can get a long first, and check its bits. This will have better performance than read byte by byte

theweipeng replied:
Great suggestion! Currently, the maximum bit length that the Number type can safely represent in JavaScript is 53 bits, which is insufficient to decode our varUInt64. BigInt serves as a workaround; however, its performance in JavaScript is significantly suboptimal. I plan to address the performance issues associated with handling 64-bit numbers comprehensively in subsequent pull requests.

chaokunyang commented on javascript/packages/fury/lib/classResolver.ts:
```diff
@@ -17,87 +17,102 @@
-    this.internalSerializer[InternalSerializerType.FURY_PRIMITIVE_DOUBLE_ARRAY] = doubleArraySerializer(fury);
-    this.internalSerializer[InternalSerializerType.BINARY] = binarySerializer(fury);
+    this.internalSerializer[InternalSerializerType.ANY] = anySerializer(fury);
+    this.regSerializer(fury, Type.string());
```
Why not name it as `registerSerializer`, `regSerializer` seems a little vague.

chaokunyang reviewed (approved):
LGTM
